core/test/preconditioner/isai.cpp

```diff
@@ -0,0 +1,139 @@
+
+
+#include <memory>
+
+
+#include <gtest/gtest.h>
+
+
+#include <ginkgo/core/base/exception.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+
+
+#include "core/test/utils.hpp"
+
+
+namespace {
+
+
+struct DummyOperator : public gko::EnableLinOp<DummyOperator>,
+                       gko::EnableCreateMethod<DummyOperator> {
+    DummyOperator(std::shared_ptr<const gko::Executor> exec,
+                  gko::dim<2> size = {})
+        : gko::EnableLinOp<DummyOperator>(exec, size)
+    {}
+
+    void apply_impl(const LinOp *b, LinOp *x) const override {}
+
+    void apply_impl(const LinOp *alpha, const LinOp *b, const LinOp *beta,
+                    LinOp *x) const override
+    {}
+};
+
+
+template <typename ValueIndexType>
+class IsaiFactory : public ::testing::Test {
+protected:
+    using value_type =
+        typename std::tuple_element<0, decltype(ValueIndexType())>::type;
+    using index_type =
+        typename std::tuple_element<1, decltype(ValueIndexType())>::type;
+    using LowerIsai = gko::preconditioner::LowerIsai<value_type, index_type>;
+    using UpperIsai = gko::preconditioner::UpperIsai<value_type, index_type>;
+    using Csr = gko::matrix::Csr<value_type, index_type>;
+
+    IsaiFactory()
+        : exec(gko::ReferenceExecutor::create()),
+          lower_isai_factory(LowerIsai::build().on(exec)),
+          upper_isai_factory(UpperIsai::build().on(exec))
+    {}
+
+    std::shared_ptr<const gko::Executor> exec;
+    std::unique_ptr<typename LowerIsai::Factory> lower_isai_factory;
+    std::unique_ptr<typename UpperIsai::Factory> upper_isai_factory;
+};
+
+TYPED_TEST_CASE(IsaiFactory, gko::test::ValueIndexTypes);
+
+
+TYPED_TEST(IsaiFactory, KnowsItsExecutor)
+{
+    ASSERT_EQ(this->lower_isai_factory->get_executor(), this->exec);
+    ASSERT_EQ(this->upper_isai_factory->get_executor(), this->exec);
+}
+
+
+TYPED_TEST(IsaiFactory, ThrowsWrongDimensionL)
+{
+    using Csr = typename TestFixture::Csr;
+    auto mtx = Csr::create(this->exec, gko::dim<2>{1, 2}, 1);
+
+    ASSERT_THROW(this->lower_isai_factory->generate(gko::share(mtx)),
+                 gko::DimensionMismatch);
+}
+
+
+TYPED_TEST(IsaiFactory, ThrowsWrongDimensionU)
+{
+    using Csr = typename TestFixture::Csr;
+    auto mtx = Csr::create(this->exec, gko::dim<2>{1, 2}, 1);
+
+    ASSERT_THROW(this->upper_isai_factory->generate(gko::share(mtx)),
+                 gko::DimensionMismatch);
+}
+
+
+TYPED_TEST(IsaiFactory, ThrowsNoConversionCsrL)
+{
+    using Csr = typename TestFixture::Csr;
+    auto mtx = DummyOperator::create(this->exec, gko::dim<2>{2, 2});
+
+    ASSERT_THROW(this->lower_isai_factory->generate(gko::share(mtx)),
+                 gko::NotSupported);
+}
+
+
+TYPED_TEST(IsaiFactory, ThrowsNoConversionCsrU)
+{
+    using Csr = typename TestFixture::Csr;
+    auto mtx = DummyOperator::create(this->exec, gko::dim<2>{2, 2});
+
+    ASSERT_THROW(this->upper_isai_factory->generate(gko::share(mtx)),
+                 gko::NotSupported);
+}
+
```

Comment:
I think you should also check if the factory parameter `skip_sorting` is being set properly.

Reply:
Good idea, I will add this test.